proposals/0000-th-stage-hygiene.rst

```diff
@@ -0,0 +1,682 @@
+  the same version of the package must be used for both.
+
+Static and dynamic linking
+  It's hard to build large swaths of the ecosystem with just static libraries. [#thanks-mboes]_
+  This is because GHCi "prefers" loading dynamic libraries.
+  The easiest thing to do is build both ways and then throw away the shared libraries.
+  The better thing to do is think of shared GHC -> static outputs as cross compilation.
+  Then only stage < 0 code for splices needs be built as shared libraries (or bytecode).
+  Plus, that builds aren't installed along side the stage 0 code, so there is no extra step to manually shave off the shared libraries.
+
+Costs and Drawbacks
+-------------------
+
+- This is a huge amount of work.
+  But I am fine chipping away it over a long period of time.
+
+- Even a temporary conflict between typed TH and this could slow typed TH's adoption.
+
+- I don't know of precedent for extensions that prevent modules from being linked together.
+
+- Most existing libraries with commonly used TH helpers (`lens`, `aeson`) have the TH in the same Cabal component but in a different module.
+  To leverage this proposal, we would have to refactor them to put those modules in a separate library component.
+  It would take decent amount of conditional code to still support old GHCs, and even more to not be a breaking change on those old libraries.
+
+Alternatives
+------------
+
+There is no fundamental reason modules couldn't export non-stage-0 items, and libraries expose non-stage-0 modules.
+At the cost of more complexity, there could be a `.lib` or `.so` for each exposed stage, and imports would be offset to match the ``#import <offset>`` literal.
+But in fairness, this might allow a smoother transition form how libraries are structured today.
+Not only would GHC need to learn more tricks, but also Cabal and other tools.
+For example, one could do ``#import 1 Control.Lens.Lens`` in ``Control.Lens.TH`` while exposing ``Control.Lens.TH`` from the same library just like today.
+I decided against this as a matter of taste.
+I think it good to enforce the normal form that the "main" stage is stage 0.
+As to the specific example, I would rather packages leverage public Cabal sub-libraries for Template Haskell anyways;
+I think that's a cleaner way to package code.
+
+It would be nice to go straight implement typed TH by quoting types.
+One would do something like::
+  [|| e... :: t... ||] :: Q (TExp [t| t... |])
+This correctly gets ``t...`` from the right for the argument to ``TExp``.
+The problem though is the parameter is type syntax rather than a type; we've thrown away all the semantics.
+We can explicitly normalize, however::
+  [|| e... :: t... ||] :: Q (TExp (Normalize [t| t... |])
+``Normalize`` could be some sort of magic type family that just tells GHC to do its thing, rather than a implementation of Haskell.
```

Comment:
```suggestion
``Normalize`` could be some sort of magic type family that just tells GHC to do its thing, rather than an implementation of Haskell.
```